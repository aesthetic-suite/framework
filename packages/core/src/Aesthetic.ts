/* eslint-disable no-param-reassign, lines-between-class-members, no-dupe-class-members */

import { FontFace as SSSFontFace, formatFontFace, LocalBlock } from '@aesthetic/sss';
import { Theme, ThemeRegistry } from '@aesthetic/system';
import {
  ClassName,
  Direction,
  Engine,
  FontFace,
  Keyframes,
  RenderOptions,
  ThemeName,
} from '@aesthetic/types';
import { arrayLoop, createState, isDOM } from '@aesthetic/utils';
import StyleSheet from './StyleSheet';
import {
  AestheticOptions,
  EventListener,
  EventType,
  GlobalSheet,
  GlobalSheetFactory,
  LocalSheet,
  LocalSheetFactory,
  OnChangeDirection,
  OnChangeTheme,
  RenderResultSheet,
  SheetParams,
} from './types';

export default class Aesthetic<Result = ClassName, Block extends object = LocalBlock> {
  protected activeDirection = createState<Direction>();

  protected activeTheme = createState<ThemeName>();

  protected globalSheetRegistry = new Map<ThemeName, GlobalSheet<unknown, Block, Result>>();

  protected listeners = new Map<EventType, Set<EventListener>>();

  protected options: AestheticOptions = {};

  protected styleEngine = createState<Engine<Result>>();

  protected themeRegistry = new ThemeRegistry<Block>();

  constructor(options?: AestheticOptions) {
    if (options) {
      this.configure(options);
    }

    // Must be bound manually because of method overloading
    this.emit = this.emit.bind(this);
    this.subscribe = this.subscribe.bind(this);
    this.unsubscribe = this.unsubscribe.bind(this);
  }

  /**
   * Change the active direction.
   */
  changeDirection = (direction: Direction, propagate: boolean = true): void => {
    if (direction === this.activeDirection.get()) {
      return;
    }

    const engine = this.getEngine();

    // Set the active direction
    this.activeDirection.set(direction);

    // Update engine direction
    engine.direction = direction;
    engine.setDirection(direction);

    // Let consumers know about the change
    if (propagate) {
      this.emit('change:direction', [direction]);
    }
  };

  /**
   * Change the active theme.
   */
  changeTheme = (name: ThemeName, propagate: boolean = true): void => {
    if (name === this.activeTheme.get()) {
      return;
    }

    const theme = this.getTheme(name);
    const engine = this.getEngine();

    // Set the active theme
    this.activeTheme.set(name);

    // Set root variables
    if (this.options.rootVariables) {
      engine.setRootVariables(theme.toVariables());
    }

    // Render theme styles
    const results = this.renderThemeStyles(theme);

    // Update engine theme
    engine.setTheme(results);

    // Let consumers know about the change
    if (propagate) {
      this.emit('change:theme', [name, results]);
    }
  };

  /**
   * Configure options unique to this instance.
   */
  configure = (customOptions: AestheticOptions): void => {
    Object.assign(this.options, customOptions);

    // Configure the engine with itself to reapply options
    const engine = this.styleEngine.get();

    if (engine) {
      this.configureEngine(engine);
    }
  };

  /**
   * Configuring which styling engine to use.
   */
  configureEngine = (engine: Engine<Result>): void => {
    const { options } = this;

    if (!engine.directionConverter && options.directionConverter) {
      engine.directionConverter = options.directionConverter;
    }

    if (!engine.unitSuffixer && options.defaultUnit) {
      engine.unitSuffixer = options.defaultUnit;
    }

    if (!engine.vendorPrefixer && options.vendorPrefixer) {
      engine.vendorPrefixer = options.vendorPrefixer;
    }

    this.styleEngine.set(engine);
  };

  /**
   * Create a local style sheet for use within components.
   */
  createComponentStyles = <T = unknown>(
    factory: LocalSheetFactory<T, Block>,
  ): LocalSheet<T, Block, Result> => new StyleSheet('local', factory);

  /**
   * Create a global style sheet for root theme styles.
   */
  createThemeStyles = <T = unknown>(
    factory: GlobalSheetFactory<T, Block>,
  ): GlobalSheet<T, Block, Result> => new StyleSheet('global', factory);

  /**
   * Emit all listeners by type, with the defined arguments.
   */
  emit(type: 'change:direction', args: Parameters<OnChangeDirection>): void;
  emit(type: 'change:theme', args: Parameters<OnChangeTheme>): void;
  emit(type: EventType, args: unknown[]): void {
    this.getListeners(type).forEach((listener) => {
      listener(...args);
    });
  }

  /**
   * Generate a class name using the selectors of a style sheet.
   * If an object is provided, it will be used to check for variants.
   */
  generateClassName = <T extends string>(
    keys: T[],
    variants: string[],
    classNames: RenderResultSheet<ClassName>,
  ): ClassName => {
    let className = '';

    arrayLoop(keys, (key) => {
      const hash = classNames[key];

      if (!hash) {
        return;
      }

      if (hash.result) {
        className += ' ';
        className += hash.result;
      }

      if (hash.variants) {
        arrayLoop(variants, (variant) => {
          if (hash.variants?.[variant]) {
            className += ' ';
            className += hash.variants[variant];
          }
        });
      }
    });

    return className.trim();
  };

  /**
   * Return the active direction for the entire application. If an active direction is undefined,
   * it will be detected from the browser's `dir` attribute.
   */
  getActiveDirection = (): Direction => {
    const active = this.activeDirection.get();

    if (active) {
      return active;
    }

    // Detect theme from engine
    const direction: Direction = this.getEngine().direction || 'ltr';

    this.changeDirection(direction);

    return direction;
  };

  /**
   * Return the currently active theme instance. If an active instance has not been defined,
   * one will be detected from the client's browser preferences.
   */
  getActiveTheme = (): Theme<Block> => {
    const active = this.activeTheme.get();

    if (active) {
      return this.getTheme(active);
    }

    // Detect theme from browser preferences
    const engine = this.getEngine();
    const theme = this.themeRegistry.getPreferredTheme({
      matchColorScheme: engine.prefersColorScheme,
      matchContrastLevel: engine.prefersContrastLevel,
    });

    this.changeTheme(theme.name);

    return theme;
  };

  /**
   * Return the current style engine.
   */
  getEngine = (): Engine<Result> => {
    const current = global.AESTHETIC_CUSTOM_ENGINE || this.styleEngine.get();

    if (current) {
      return current;
    }

    throw new Error('No style engine defined. Have you configured one with `configureEngine()`?');
  };

  /**
   * Return a set of listeners, or create it if it does not exist.
   */
  getListeners = <T extends Function>(type: EventType): Set<T> => {
    const set = this.listeners.get(type) || new Set();

    this.listeners.set(type, set);

    return (set as unknown) as Set<T>;
  };

  /**
   * Return a theme instance by name.
   */
  getTheme = (name: ThemeName): Theme<Block> => this.themeRegistry.getTheme(name);

  /**
   * Register a theme, with optional global theme styles.
   */
  registerTheme = (
    name: ThemeName,
    theme: Theme<Block>,
    sheet: GlobalSheet<unknown, Block, Result> | null = null,
    isDefault: boolean = false,
  ): void => {
    this.themeRegistry.register(name, theme, isDefault);

    if (sheet) {
      if (__DEV__) {
        if (!(sheet instanceof StyleSheet) || sheet.type !== 'global') {
          throw new TypeError('Rendering theme styles require a `GlobalSheet` instance.');
        }
      }

      this.globalSheetRegistry.set(name, sheet);
    }
  };

  /**
   * Register a default light or dark theme, with optional global theme styles.
   */
  registerDefaultTheme = (
    name: ThemeName,
    theme: Theme<Block>,
    sheet: GlobalSheet<unknown, Block, Result> | null = null,
  ): void => {
    this.registerTheme(name, theme, sheet, true);
  };

  /**
   * Render a component style sheet to the document with the defined style query parameters.
   */
  renderComponentStyles = <T = unknown>(
    sheet: LocalSheet<T, Block, Result>,
    themeName?: ThemeName,
  ) => {
    if (__DEV__) {
      if (!(sheet instanceof StyleSheet) || sheet.type !== 'local') {
        throw new TypeError('Rendering component styles require a `LocalSheet` instance.');
      }
    }

    const theme = themeName ? this.getTheme(themeName) : this.getActiveTheme();

    return sheet.render(this.getEngine(), theme, {
      customProperties: this.options.customProperties,
      deterministic: this.options.deterministicClasses,
      direction: this.getActiveDirection(),
      vendor: !!this.options.vendorPrefixer,
    });
  };

  /**
   * Render a `@font-face` to the global style sheet and return the font family name.
   */
  renderFontFace = (
    fontFace: FontFace | SSSFontFace,
    fontFamily?: string,
    params?: RenderOptions,
  ): string =>
    this.getEngine().renderFontFace(
      formatFontFace({
        fontFamily,
        ...fontFace,
      }) as FontFace,
      params,
    );

  /**
   * Render an `@import` to the global style sheet and return the import path.
   */
  renderImport = (path: string, params?: RenderOptions): string =>
    this.getEngine().renderImport(path, params);

  /**
   * Render a `@keyframes` to the global style sheet and return the animation name.
   */
  renderKeyframes = (
    keyframes: Keyframes,
    animationName?: string,
    params?: RenderOptions,
  ): string => this.getEngine().renderKeyframes(keyframes, animationName, params);

  /**
   * Render a global theme style sheet and return a result, if one was generated.
   */
  renderThemeStyles = (theme: Theme<Block>, params: SheetParams = {}): Result[] => {
    const sheet = this.globalSheetRegistry.get(theme.name);
    const results: Result[] = [];

    // Render theme CSS variables
    if (isDOM()) {
      results.push(this.getEngine().renderRuleGrouped(theme.toVariables(), { type: 'global' }));
    }

    // Render theme styles
    if (sheet) {
      const result = sheet.render(this.getEngine(), theme, {
        customProperties: this.options.customProperties,
        direction: this.getActiveDirection(),
        vendor: !!this.options.vendorPrefixer,
        ...params,
      })['@root']?.result;

      if (result) {
        results.push(result);
      }
    }

    return results;
  };

  /**
   * Subscribe and listen to an event by name.
   */
  subscribe(type: 'change:direction', listener: OnChangeDirection): () => void;
  subscribe(type: 'change:theme', listener: OnChangeTheme): () => void;
  subscribe(type: EventType, listener: Function): () => void {
    this.getListeners(type).add(listener);

    return () => {
      this.unsubscribe(type as 'change:theme', listener as OnChangeTheme);
    };
  }

  /**
   * Unsubscribe from an event by name.
   */
  unsubscribe(type: 'change:direction', listener: OnChangeDirection): void;
  unsubscribe(type: 'change:theme', listener: OnChangeTheme): void;
  unsubscribe(type: EventType, listener: Function): void {
    this.getListeners(type).delete(listener);
  }
}
