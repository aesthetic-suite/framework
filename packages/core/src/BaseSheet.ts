export default abstract class BaseSheet {
  protected validateFactory<T>(factory: T): T {
    if (__DEV__) {
      const typeOf = typeof factory;

      if (typeOf !== 'function') {
        throw new TypeError(`A style sheet factory function is required, found "${typeOf}".`);
      }
    }

    return factory;
  }
}
