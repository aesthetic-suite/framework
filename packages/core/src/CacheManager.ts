import { Direction, ThemeName } from './types';

export interface CacheTags {
  dir?: Direction;
  global?: boolean;
  theme?: ThemeName;
}

export interface CacheUnit<T> extends CacheTags {
  value: T;
}

export default class CacheManager<T> {
  protected cache: Map<string, CacheUnit<T>[]> = new Map();

  compare(unit: CacheUnit<T>, tags: CacheTags): boolean {
    return unit.dir === tags.dir && unit.global === tags.global && unit.theme === tags.theme;
  }

  get(key: string, tags: CacheTags): T | null {
    const units = this.cache.get(key);

    if (!units || units.length === 0) {
      return null;
    }

    const cache = units.find(unit => this.compare(unit, tags));

    return cache ? cache.value : null;
  }

  set(key: string, value: T, tags: CacheTags): T {
    const units = this.cache.get(key) || [];

    units.push({
      ...tags,
      value,
    });

    this.cache.set(key, units);

    return value;
  }
}
