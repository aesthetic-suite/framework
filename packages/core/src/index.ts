/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import Aesthetic, { AestheticOptions } from './Aesthetic';
import ClassNameAesthetic from './ClassNameAesthetic';
import UnifiedSyntax from './UnifiedSyntax';
import Ruleset from './Ruleset';
import Sheet from './Sheet';
import injectCSS from './helpers/injectCSS';

export { AestheticOptions, ClassNameAesthetic, UnifiedSyntax, Ruleset, Sheet, injectCSS };

export * from './types';

export default Aesthetic;
