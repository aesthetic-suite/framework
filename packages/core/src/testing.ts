import { getRenderedStyles, purgeStyles, purgeStyles } from '@aesthetic/style/lib/testing';
import { design, lightTheme, darkTheme } from '@aesthetic/system/lib/testing';
import Aesthetic from './Aesthetic';

export { design, lightTheme, darkTheme, getRenderedStyles, purgeStyles, purgeStyles };

export function setupAesthetic(aesthetic: Aesthetic) {
  aesthetic.registerDefaultTheme('day', lightTheme);
  aesthetic.registerDefaultTheme('night', darkTheme);
}

export function teardownAesthetic(aesthetic: Aesthetic) {
  lightTheme.name = '';
  darkTheme.name = '';
  aesthetic.resetForTesting();
}
