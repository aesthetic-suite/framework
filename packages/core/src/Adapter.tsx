import { isObject, stripClassPrefix } from 'aesthetic-utils';
import aesthetic from './instance';
import CacheManager from './CacheManager';
import Sheet from './Sheet';
import UnifiedSyntax from './UnifiedSyntax';
import { ClassName, CompiledStyleSheet, StyleName, TransformOptions } from './types';

export default abstract class Adapter<NativeBlock extends object> {
  protected cacheManager = new CacheManager<CompiledStyleSheet>();

  protected syntax = new UnifiedSyntax<NativeBlock>();

  adaptStyleSheet(nativeSheet: Sheet<NativeBlock>, styleName: StyleName): CompiledStyleSheet {
    return nativeSheet.toObject();
  }

  compileStyleSheet(styleName: StyleName, baseOptions?: TransformOptions): CompiledStyleSheet {
    const options = this.prepareTransformOptions(baseOptions);
    const cache = this.cacheManager.get(styleName, options);

    if (cache) {
      return cache;
    }

    // Apply global styles on first render
    this.applyGlobalStyles(baseOptions);

    // Convert unified syntax to adapter native syntax
    const nativeSheet = this.syntax.convertStyleSheet(
      aesthetic.getStyleSheet(styleName, options.theme),
      {
        ...options,
        name: styleName,
      },
    );

    // Parse and adapt native syntax
    const adaptedSheet = this.adaptStyleSheet(nativeSheet, styleName);

    return this.cacheManager.set(
      styleName,
      {
        ...adaptedSheet,
        ...nativeSheet.classNames,
      },
      options,
    );
  }

  /**
   * Transform the list of class names or style objects to a single class name.
   */
  transformStyles(styles: (undefined | false | ClassName | object)[]): ClassName {
    const classNames: ClassName[] = [];
    const objects: object[] = [];

    styles.forEach(style => {
      if (!style) {
        return;
      }

      if (typeof style === 'string') {
        classNames.push(
          ...String(style)
            .split(' ')
            .map(s => stripClassPrefix(s).trim()),
        );
      } else if (isObject(style)) {
        objects.push(style);
      } else if (__DEV__) {
        throw new Error('Unsupported style type to transform.');
      }
    });

    // Transform parsed blocks to class names
    if (objects.length > 0) {
      classNames.push(this.transformToClassName(objects));
    }

    return classNames.join(' ');
  }

  /**
   * Transform the compiled style objects into a class name.
   */
  abstract transformToClassName(styles: object[]): ClassName;

  /**
   * Return transform options with defaults applied.
   */
  protected prepareTransformOptions(options: TransformOptions = {}): Required<TransformOptions> {
    const dir = aesthetic.options.rtl ? 'rtl' : 'ltr';

    return {
      dir: options.dir || dir,
      global: options.global || false,
      name: options.name || '',
      theme: options.theme || aesthetic.options.theme,
    };
  }
}
