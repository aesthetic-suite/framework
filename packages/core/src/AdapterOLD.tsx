import deepMerge from 'extend';
import uuid from 'uuid/v4';
import { isObject, stripClassPrefix } from 'aesthetic-utils';
import CacheManager from './CacheManager';
import Sheet from './Sheet';
import StyleSheetManager from './StyleSheetManager';
import UnifiedSyntax from './UnifiedSyntax';
import { GLOBAL_STYLE_NAME } from './constants';
import {
  ClassName,
  TransformOptions,
  SheetMap,
  StyleName,
  StyleSheetFactory,
  ThemeName,
} from './types';

export default abstract class Adapter<
  Theme extends object,
  NativeBlock extends object,
  ParsedBlock extends object | string = NativeBlock
> {
  syntax = new UnifiedSyntax<NativeBlock>();

  protected sheetManager: StyleSheetManager | null = null;

  /**
   * Apply and inject global styles for the current theme.
   * This should only happen once!
   */
  applyGlobalStyles(baseOptions?: TransformOptions): this {
    // Templates may be generated by Webpack or something similar,
    // so set this programmatically based on the defined option.
    if (typeof document !== 'undefined') {
      document.documentElement.setAttribute('dir', this.options.rtl ? 'rtl' : 'ltr');
    }

    const options = this.getPreparedTransformOptions({
      ...baseOptions,
      global: true,
      name: GLOBAL_STYLE_NAME,
    });

    // Direction changes shouldn't regenerate global styles
    delete options.dir;

    const cache = this.cacheManager.get(GLOBAL_STYLE_NAME, options);
    const globalDef = this.globals[options.theme];

    if (cache || !globalDef) {
      return this;
    }

    const globalSheet = globalDef(this.getTheme(options.theme));
    const parsedSheet = this.cacheManager.set(
      GLOBAL_STYLE_NAME,
      this.parseStyleSheet(
        this.syntax.convertGlobalSheet(globalSheet, options).toObject(),
        GLOBAL_STYLE_NAME,
      ),
      options,
    );

    // Some adapters require the styles to be transformed to be flushed
    this.transformStyles(Object.values(parsedSheet), options);
    this.flushStyles(GLOBAL_STYLE_NAME);

    return this;
  }

  /**
   * Change the current theme to another registered theme.
   * This will purge all flushed global styles and regenerate new ones.
   */
  changeTheme(themeName: ThemeName): this {
    const oldTheme = this.options.theme;

    // Set theme as new option
    this.getTheme(themeName);
    this.options.theme = themeName;

    // Purge previous global styles
    this.purgeStyles(GLOBAL_STYLE_NAME);
    this.cacheManager.clear(unit => unit.global === true && unit.theme === oldTheme);

    // Generate new global styles
    this.applyGlobalStyles({ theme: themeName });

    return this;
  }

  /**
   * Create and return a style sheet unique to an adapter.
   */
  createStyleSheet(styleName: StyleName, baseOptions?: TransformOptions): SheetMap<ParsedBlock> {
    // Apply global styles on first render
    this.applyGlobalStyles(baseOptions);

    const nativeSheet = this.syntax.convertStyleSheet(
      this.getStyleSheet(styleName, options.theme),
      {
        ...options,
        name: styleName,
      },
    );

    const parsedSheet = this.parseStyleSheet(nativeSheet.toObject(), styleName);

    return this.cacheManager.set(
      styleName,
      {
        ...parsedSheet,
        ...nativeSheet.classNames,
      } as SheetMap<ParsedBlock>,
      options,
    );
  }

  /**
   * Compose and extend multiple style sheets to create 1 style sheet.
   */
  extendStyles(...styleSheets: StyleSheetFactory<Theme, any>[]): StyleSheetFactory<Theme, any> {
    return (theme: Theme) => {
      const sheets = styleSheets.map(sheet => sheet(theme));

      return deepMerge(true, {}, ...sheets);
    };
  }

  /**
   * Flush a target component's transformed styles and inject them into the DOM.
   * If no target defined, will flush all buffered styles.
   */
  flushStyles(styleName?: StyleName) {}

  /**
   * Return true if the style object is a parsed block and not a native block.
   */
  isParsedBlock(block: NativeBlock | ParsedBlock): block is ParsedBlock {
    return isObject(block);
  }

  /**
   * Parse an Aesthetic style sheet into an adapter native style sheet.
   */
  parseStyleSheet(styleSheet: SheetMap<NativeBlock>, styleName: StyleName): SheetMap<ParsedBlock> {
    // @ts-ignore Allow spread
    return { ...styleSheet };
  }

  /**
   * Purge and remove all styles from the DOM for the target component.
   * If no target defined, will purge all possible styles.
   */
  purgeStyles(styleName?: StyleName) {}

  /**
   * Return a native style sheet manager used for injecting CSS.
   */
  protected getStyleSheetManager(): StyleSheetManager {
    if (this.sheetManager) {
      return this.sheetManager;
    }

    this.sheetManager = new StyleSheetManager();

    return this.sheetManager;
  }
}
