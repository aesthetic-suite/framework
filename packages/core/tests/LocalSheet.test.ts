import converter from '@aesthetic/addon-direction';
import { ClientRenderer } from '@aesthetic/style';
import { StyleSheet, LocalSheet, createComponentStyles } from '../src';
import { getRenderedStyles, lightTheme, purgeStyles } from '../src/testing';

describe('LocalSheet', () => {
  let renderer: ClientRenderer;
  let sheet: LocalSheet;

  beforeEach(() => {
    renderer = new ClientRenderer({ converter });
    sheet = createComponentStyles(() => ({
      foo: {
        display: 'block',
        background: 'white',
        color: 'black',
        textAlign: 'left',
        fontSize: 12,
        fontFamily: [
          '"Open Sans"',
          {
            fontFamily: 'Roboto',
            fontStyle: 'normal',
            fontWeight: 'normal',
            srcPaths: ['fonts/Roboto.woff2', 'fonts/Roboto.ttf'],
          },
        ],
        ':hover': {
          color: 'red',
        },
        '@selectors': {
          ':focus': {
            outline: 'red',
          },
        },
        '@media': {
          '(max-width: 1000px)': {
            display: 'none',
            borderWidth: 1,
          },
        },
      },
      bar: {
        transition: '200ms all',
        animationName: {
          from: { opacity: 0 },
          to: { opacity: 1 },
        },
      },
      baz: 'class-baz',
    }));
  });

  afterEach(() => {
    purgeStyles();
  });

  it('errors if a non-function factory is passed', () => {
    expect(
      () =>
        new StyleSheet(
          'local',
          // @ts-expect-error
          123,
        ),
    ).toThrow('A style sheet factory function is required, found "number".');
  });

  it('only renders once when cached', () => {
    const spy = jest.spyOn(sheet, 'compose');

    sheet.render(renderer, lightTheme, {});
    sheet.render(renderer, lightTheme, {});
    sheet.render(renderer, lightTheme, {});

    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('re-renders when params change', () => {
    const spy = jest.spyOn(sheet, 'compose');

    sheet.render(renderer, lightTheme, {});
    sheet.render(renderer, lightTheme, { direction: 'rtl' });
    sheet.render(renderer, lightTheme, {});
    sheet.render(renderer, lightTheme, { direction: 'rtl' });

    expect(spy).toHaveBeenCalledTimes(2);
    expect(getRenderedStyles('standard')).toMatchSnapshot();
  });

  it('renders and returns an object of class names', () => {
    const classes = sheet.render(renderer, lightTheme, {});

    expect(classes).toEqual({
      foo: { class: 'a b c d e f g h i j' },
      bar: { class: 'k l' },
      baz: { class: 'class-baz' },
    });
    expect(getRenderedStyles('standard')).toMatchSnapshot();
  });

  it('renders a declaration for each rule property', () => {
    const spy = jest.spyOn(renderer, 'renderDeclaration');

    sheet.render(renderer, lightTheme, {});

    expect(spy).toHaveBeenCalledWith('display', 'block', expect.any(Object));
    expect(spy).toHaveBeenCalledWith('background', 'white', expect.any(Object));
    expect(spy).toHaveBeenCalledWith('color', 'black', expect.any(Object));
    expect(spy).toHaveBeenCalledWith('fontSize', 12, expect.any(Object));
    expect(spy).toHaveBeenCalledWith('fontFamily', '"Open Sans", Roboto', expect.any(Object));
  });

  it('renders @font-face', () => {
    const spy = jest.spyOn(renderer, 'renderFontFace');

    sheet.render(renderer, lightTheme, {});

    expect(spy).toHaveBeenCalledWith(
      {
        fontFamily: 'Roboto',
        fontStyle: 'normal',
        fontWeight: 'normal',
        src:
          "url('fonts/Roboto.woff2') format('woff2'), url('fonts/Roboto.ttf') format('truetype')",
      },
      {
        direction: 'ltr',
        rankings: expect.any(Object),
        unit: 'px',
        vendor: false,
      },
    );
    expect(getRenderedStyles('global')).toMatchSnapshot();
  });

  it('renders @keyframes', () => {
    const spy = jest.spyOn(renderer, 'renderKeyframes');

    sheet.render(renderer, lightTheme, {});

    expect(spy).toHaveBeenCalledWith(
      {
        from: { opacity: 0 },
        to: { opacity: 1 },
      },
      '',
      {
        direction: 'ltr',
        rankings: expect.any(Object),
        unit: 'px',
        vendor: false,
      },
    );
    expect(getRenderedStyles('global')).toMatchSnapshot();
  });

  describe('variants', () => {
    beforeEach(() => {
      sheet.addColorSchemeVariant('dark', () => ({
        foo: {
          background: 'black',
          color: 'white',
        },
      }));

      sheet.addContrastVariant('high', () => ({
        foo: {
          background: 'pink',
        },
      }));

      sheet.addContrastVariant('low', () => ({
        foo: {
          background: 'yellow',
        },
      }));

      sheet.addThemeVariant('danger', () => ({
        foo: {
          background: 'red',
          color: 'yellow',
        },
      }));
    });

    it('errors for invalid color scheme name', () => {
      expect(() => {
        sheet.addColorSchemeVariant(
          // @ts-expect-error
          'unknown',
          () => ({}),
        );
      }).toThrow('Color scheme variant must be one of "light" or "dark".');
    });

    it('errors for invalid contrast name', () => {
      expect(() => {
        sheet.addContrastVariant(
          // @ts-expect-error
          'unknown',
          () => ({}),
        );
      }).toThrow('Contrast level variant must be one of "high", "low", or "normal".');
    });

    it('inherits color scheme', () => {
      const spy = jest.spyOn(renderer, 'renderDeclaration');

      sheet.render(renderer, lightTheme, {
        scheme: 'dark',
      });

      expect(spy).toHaveBeenCalledWith('background', 'black', expect.any(Object));
      expect(spy).toHaveBeenCalledWith('color', 'white', expect.any(Object));
    });

    it('inherits high contrast', () => {
      const spy = jest.spyOn(renderer, 'renderDeclaration');

      sheet.render(renderer, lightTheme, {
        contrast: 'high',
      });

      expect(spy).toHaveBeenCalledWith('background', 'pink', expect.any(Object));
      expect(spy).toHaveBeenCalledWith('color', 'black', expect.any(Object));
    });

    it('inherits low contrast', () => {
      const spy = jest.spyOn(renderer, 'renderDeclaration');

      sheet.render(renderer, lightTheme, {
        contrast: 'low',
      });

      expect(spy).toHaveBeenCalledWith('background', 'yellow', expect.any(Object));
      expect(spy).toHaveBeenCalledWith('color', 'black', expect.any(Object));
    });

    it('inherits theme by name', () => {
      const spy = jest.spyOn(renderer, 'renderDeclaration');

      sheet.render(renderer, lightTheme, {
        theme: 'danger',
      });

      expect(spy).toHaveBeenCalledWith('background', 'red', expect.any(Object));
      expect(spy).toHaveBeenCalledWith('color', 'yellow', expect.any(Object));
    });

    it('contrast overrides scheme', () => {
      const spy = jest.spyOn(renderer, 'renderDeclaration');

      sheet.render(renderer, lightTheme, {
        scheme: 'dark',
        contrast: 'low',
      });

      expect(spy).toHaveBeenCalledWith('background', 'yellow', expect.any(Object));
      expect(spy).toHaveBeenCalledWith('color', 'white', expect.any(Object));
    });

    it('theme overrides contrast and scheme', () => {
      const spy = jest.spyOn(renderer, 'renderDeclaration');

      sheet.render(renderer, lightTheme, {
        scheme: 'dark',
        contrast: 'high',
        theme: 'danger',
      });

      expect(spy).toHaveBeenCalledWith('background', 'red', expect.any(Object));
      expect(spy).toHaveBeenCalledWith('color', 'yellow', expect.any(Object));
    });
  });
});
