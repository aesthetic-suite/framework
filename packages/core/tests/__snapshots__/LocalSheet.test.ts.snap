// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocalSheet re-renders when params change 1`] = `".a {display: block;}.b {background: white;}.c {color: black;}.d {text-align: left;}.e {font-size: 12px;}.f {font-family: \\"Open Sans\\", Roboto;}.g:hover {color: red;}.j:focus {outline: red;}.k {transition: 200ms all;}.l {animation-name: kf1plt5bd;}.m {text-align: right;}"`;

exports[`LocalSheet renders @font-face 1`] = `
"@font-face {font-family: Roboto; font-style: normal; font-weight: normal; src: url('fonts/Roboto.woff2') format('woff2'), url('fonts/Roboto.ttf') format('truetype');}@keyframes kf1plt5bd { 
  from {opacity: 0;} 
  to {opacity: 1;} 
}"
`;

exports[`LocalSheet renders @keyframes 1`] = `
"@font-face {font-family: Roboto; font-style: normal; font-weight: normal; src: url('fonts/Roboto.woff2') format('woff2'), url('fonts/Roboto.ttf') format('truetype');}@keyframes kf1plt5bd { 
  from {opacity: 0;} 
  to {opacity: 1;} 
}"
`;

exports[`LocalSheet renders and returns an object of class names 1`] = `".a {display: block;}.b {background: white;}.c {color: black;}.d {text-align: left;}.e {font-size: 12px;}.f {font-family: \\"Open Sans\\", Roboto;}.g:hover {color: red;}.j:focus {outline: red;}.k {transition: 200ms all;}.l {animation-name: kf1plt5bd;}"`;
